### PR TITLE
ARM targeting support

### PR DESCRIPTION
**Pull request by sr229**
This PR is part of a progress to build `code-server` for ARM in Linux. (cdr/code-server#35)

Currently the build supports Alpine but may extend to Debian later on (CentOS is not supported atm)

### CODE REVIEW

File: aarch64-alpine.dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM balenalib/aarch64-alpine-node:10.15-edge-build
+
+RUN ["cross-build-start"]
+
+RUN apk update && apk add bash gcc g++ ccache git make python linux-headers
```
Comment (SuperSandro2000) [inline aarch64-alpine.dockerfile]: You don't do ``apk update``.
You should probably add ``--no-cache --no-progress`` to ``apk add``.

File: aarch64.dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+FROM balenalib/aarch64-debian-node:10.15-jessie-build
+
+RUN ["cross-build-start"]
+
+RUN apt-get update && apt-get -y install build-essential linux-headers-generic gcc g++ ccache git make
```
Comment (SuperSandro2000) [inline aarch64.dockerfile]: You should probably add ``install --no-install-recommends -qy``.

File: aarch64.dockerfile [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+
+RUN ["cross-build-end"]
```
Comment (SuperSandro2000) [inline aarch64.dockerfile]: Add a newline.

File: aarch64-alpine.dockerfile [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+
+RUN ["cross-build-end"]
```
Comment (SuperSandro2000) [inline aarch64-alpine.dockerfile]: Add a newline.

File: alpine.dockerfile [hunk 5]
```diff
@@ -2,12 +2,4 @@
...
-RUN apk update
-RUN apk add bash
-RUN apk add gcc
-RUN apk add g++
-RUN apk add ccache
-RUN apk add git
-RUN apk add make
-RUN apk add python
-RUN apk add linux-headers
+RUN apk update && apk add bash gcc g++ ccache git make python linux-headers
```
Comment (SuperSandro2000) [inline alpine.dockerfile]: You don't do ``apk update``.
You should probably add ``--no-cache --no-progress`` to ``apk add``.

File: armhf-alpine.dockerfile [hunk 6]
```diff
@@ -0,0 +1,7 @@
+FROM balenalib/armv7hf-alpine-node:10.15-edge-build
+
+RUN ["cross-build-start"]
+
+RUN apk update && apk add bash gcc g++ ccache git make python linux-headers
```
Comment (SuperSandro2000) [inline armhf-alpine.dockerfile]: You don't do ``apk update``.
You should probably add ``--no-cache --no-progress`` to ``apk add``.

File: armhf.dockerfile [hunk 7]
```diff
@@ -0,0 +1,7 @@
+FROM balenalib/armv7hf-debian-node:10.15-jessie-build
+
+RUN ["cross-build-start"]
+
+RUN apt-get update && apt-get -y install build-essential linux-headers-generic gcc g++ ccache git make
```
Comment (SuperSandro2000) [inline armhf.dockerfile]: You should probably add ``install --no-install-recommends -qy``.

File: scripts/docker_build.sh [hunk 8]
```diff
@@ -11,19 +11,40 @@ source ./vars.sh
...
+	 docker exec $containerID mkdir /src
+
+	 function exec() {
+	 	 docker exec $containerID bash -c "$@"
+	 }
+
+	 docker cp ../. $containerID:/src
+	 exec "cross-build-start"
+	 exec "$PREBUILD_COMMAND/src/lib/node/build.sh"
+	 exec "cd /src && npm rebuild"
```
Comment (SuperSandro2000) [inline scripts/docker_build.sh]: Why not combine this line with the next?

File: scripts/docker_build.sh [hunk 9]
```diff
@@ -11,19 +11,40 @@ source ./vars.sh
...
+	 docker start $containerID
+	 docker exec $containerID mkdir /src
+
+	 function exec() {
+	 	 docker exec $containerID bash -c "$@"
+	 }
+
+	 docker cp ../. $containerID:/src
+	 exec "$PREBUILD_COMMAND/src/lib/node/build.sh"
+	 exec "cd /src && npm rebuild"
```
Comment (SuperSandro2000) [inline scripts/docker_build.sh]: Why not combine this line with the next?

File: armhf.dockerfile (hunk 7 above)
Comment (sr229) [reply armhf.dockerfile]: -q is a quiet flag, we're building this on a CI, so what's the point of the quiet flag

File: scripts/docker_build.sh (hunk 8 above)
Comment (sr229) [reply scripts/docker_build.sh]: I don't want to change much of the original codebase, I don't think this needs moving.

File: aarch64.dockerfile (hunk 2 above)
Comment (sr229) [reply aarch64.dockerfile]: Building via CI, not needed.

File: scripts/docker_build.sh (hunk 8 above)
Comment (SuperSandro2000) [reply scripts/docker_build.sh]: Agreed.

File: armhf.dockerfile (hunk 7 above)
Comment (SuperSandro2000) [reply armhf.dockerfile]: I find it quiet annoying to debug when the CI fails and I have 1k lines that don't mean anything. Additionally  to that Travis limits the log size to about 4MB and if we exceed that it gets cut off. Source https://docs.travis-ci.com/user/common-build-problems/#log-length-exceeded

So I my opinion it is good practice to silent useless commands or try to shorten them.
``-q`` hides already some not needed output and in my personal images I cranked it up to ``-qq`` to not flood my logs.

File: scripts/docker_build.sh (hunk 9 above)
Comment (sr229) [reply scripts/docker_build.sh]: I'll see what I can do but I don't really want to move stuff if it causes something unexpected.

**Comment by sr229** [comment]
Update: CentOS is currently NOT supported due to available balenalib images being Debian, Ubuntu and Alpine based. You're on your own CentOS users.

**Comment by sr229** [comment]
CC @coadler for review ty

**Comment by coadler** [comment]
cc @code-asher @kylecarbs

**Pull request merged**